# Chapter 4 - Classification: Answers to the Review Questions

**Datasets used in the coding exercises:**
- `hr_employee_data.xlsx` (Exercise 13) - the same HR dataset used for the EDA in Chapter 3.
- MNIST handwritten digits (Exercise 15) - since scikit-learn's `fetch_openml("mnist_784")`
  (used in the book's own Code Example 1) could not reach openml.org's API from this
  environment, the identical MNIST data (same 70,000 images, same 28x28 grayscale format,
  same labels) was instead obtained as CSV from the public GitHub mirror
  `phoebetronic/mnist` - the same widely-used raw MNIST split, just distributed as a
  ready-made CSV instead of through the OpenML API.


### 1. What characterizes classification problems? Give some examples of application areas.

The book defines classification problems in Section 4.1 (p. 151):

> "Classification problems are about predicting categorical data, where the data can take
> two or more classes."

So what characterizes a classification problem is that **the target variable is
categorical** (it belongs to one of a fixed set of classes), as opposed to regression
problems (Chapter 3) where the target variable is continuous.

**Examples of application areas (p. 151-152, the book's own examples)**

- **Customer analysis** - Predicting whether a customer will *churn* (i.e. leave the
  company) or not, based on variables such as number of purchases, age, gender and number
  of support tickets.
- **Healthcare** - Predicting whether a patient has a certain disease or not (sick or
  healthy) based on medical data such as blood pressure, pulse, etc.
- **Image recognition** - Identifying objects in an image based on variables such as
  color, shape and texture - for example identifying pest damage in agricultural fields or
  on golf courses.
- **Anomaly detection** - Used, among other things, in manufacturing processes to
  automatically identify defective products using images from cameras.
- **Finance** - Predicting whether a stock will rise in value or not, based on e.g. company
  information such as profitability and revenue, and macroeconomic data such as inflation
  and interest rates.
- **Credit risk** - Assessing whether a bank customer should be granted a mortgage or not,
  based on data such as income, age, marital status and any payment remarks.

**Types of classification (p. 152-153)**

The book distinguishes three types:
- **Binary classification** - exactly two classes, e.g. {churn, stay} or {1, 0}.
- **Multiclass classification** - more than two classes, e.g. predicting which animal is
  shown in an image out of {dog, cat, horse, bird}.
- **Multioutput (multiclass) classification** - predicting two or more output variables at
  the same time, e.g. simultaneously predicting whether a person is over 18 {yes, no} and
  which gender they are {man, woman} (multioutput), or, if at least one output can take
  more than two classes (e.g. also predicting which of {southern Sweden, central Sweden,
  northern Sweden} the person lives in), *multioutput multiclass classification*. The book
  notes it does not go deeper into these last two types and focuses on binary and
  multiclass classification.


### 2. Explain how the OvR and OvO algorithms work.

The book explains this in Section 4.1.1, "OvR- and OvO-algoritmerna" (p. 154):

> "Binary classification models can, with the help of two algorithms called One-vs-Rest
> (OvR) and One-vs-One (OvO), be extended to also be able to handle multiclass
> classification."

**One-vs-Rest (OvR)**

> "OvR means that for each class in a multiclass problem, a model is trained to be able to
> distinguish whether a data point belongs to that class or not. When a new observation is
> to be classified, all the models are run, and the final prediction is the one from the
> model with the highest *score* (such as probability)."

**One-vs-One (OvO)**

> "With OvO we instead train one model per pair of classes. This means that if there are N
> classes, (N × (N-1))/2 models need to be trained. Each model compares a pair of classes
> and predicts one of the two classes. The class predicted the most times overall is
> finally chosen."



### 3. Explain the following evaluation metrics.


#### a) Confusion matrix (Section 4.2.1, p. 155)

> "A confusion matrix is a matrix that visualizes how well a model performs by comparing
> true and predicted values. Each row represents the true class, while each column
> represents the predicted class (it can also be the other way around). The number of
> times each class has been predicted as each respective class by the model is calculated
> and written into the matrix. The diagonal, from the top-left corner to the bottom-right
> corner, shows all the data points that were correctly classified."

For a **binary** classification problem, the matrix has four cells (Figure 4.2, p. 155):
**True Positive (TP)** - the number of times the model correctly predicted a positive
class; **False Negative (FN)** - the number of times the model incorrectly predicted a
negative class (when it was actually positive); **False Positive (FP)** - the number of
times the model incorrectly predicted a positive class (when it was actually negative);
**True Negative (TN)** - the number of times the model correctly predicted a negative
class. For **multiclass** problems (Figure 4.3, p. 156), the matrix extends to one
row/column per class, and the book demonstrates it in scikit-learn with
`sklearn.metrics.confusion_matrix` and visualizes it with `ConfusionMatrixDisplay` (p. 157).

#### b) Accuracy (Section 4.2.2, p. 158)

> "Accuracy is a measure of how large a share of all observations have been classified
> correctly."

Formula (Equation 4.1):

$$\text{Accuracy} = \frac{TP+TN}{TP+TN+FP+FN}$$

The book warns, however (p. 158-159): "A high value for accuracy thus means that a large
share of the observations have been classified correctly. Does this mean our model
performs well? Not necessarily. If our dataset is imbalanced, i.e. one class is
considerably more common than the others, the metric risks being misleading. If a dataset
has data points where 95% belong to class 1 and our model always predicts class 1, the
accuracy score will be 95%, despite the model not being able to predict class 0 at all."
Computed in scikit-learn with `sklearn.metrics.accuracy_score`.

#### c) Precision (Section 4.2.3, p. 159)

> "Precision measures the share of the positive predictions that are actually correct."

Formula (Equation 4.2):

$$\text{Precision} = \frac{TP}{TP+FP}$$

> "Theoretically speaking, if a high value for precision is desired, it is possible to
> build a model that only classifies an observation as positive if it is very certain.
> This will, however, generally often lead to many values that are actually positive being
> incorrectly predicted as negative. This then leads to recall [...] becoming low."

#### d) Recall (Section 4.2.3, p. 159-160)

> "Recall measures the share of the positive class that is predicted correctly."

Formula (Equation 4.3):

$$\text{Recall} = \frac{TP}{TP+FN}$$

> "Theoretically speaking, if we build a model that always classifies an observation as
> positive, we will get 100% recall (assuming there is at least one positive observation
> in the dataset). This does, however, generally lead to precision decreasing, since the
> share of our positive predictions that are actually correct decreases (we always predict
> positive regardless of whether the observation is actually positive or negative)."

Computed in scikit-learn with `sklearn.metrics.precision_score` and
`sklearn.metrics.recall_score` (with `average="macro"` for multiclass problems, i.e.
computing the metric per class and then averaging).

#### e) F1-score (Section 4.2.4, p. 162)

> "F1-score is what is called the harmonic mean of precision and recall."

Formula (Equation 4.4):

$$F_1 = \frac{2}{\text{recall}^{-1}+\text{precision}^{-1}} = 2\cdot\frac{\text{precision}\cdot\text{recall}}{\text{precision}+\text{recall}} = \frac{2TP}{2TP+FP+FN}$$

> "To get a high value for F1-score, both precision and recall need to be high."

The book demonstrates this with a worked table of precision/recall/F1 combinations showing
that F1 is low whenever either precision or recall is low (e.g. precision=1.00,
recall=0.01 gives F1=0.02), confirming that F1 rewards a *balance* between the two rather
than either alone. Computed in scikit-learn with `sklearn.metrics.f1_score`, or all three
metrics at once with `sklearn.metrics.classification_report` (see Reasoning Question 11
below for a full worked example straight from the book).

#### f) The ROC curve (Section 4.2.5, p. 164-166)

> "The ROC curve (from the English *receiver operating characteristic curve*) is similar to
> the precision-recall curve. But instead of visualizing the relationship between precision
> and recall, it visualizes the relationship between TPR, from the English *True Positive
> Rate*, and FPR, from the English *False Positive Rate*."

$$\text{TPR} = \frac{TP}{TP+FN} = \text{recall} \qquad (4.5) \qquad\qquad \text{FPR} = \frac{FP}{FP+TN} \qquad (4.6)$$

Note that TPR is simply another name for recall. The book also introduces *sensitivity*
(= recall/TPR, how well the model finds positive cases) and *specificity* (how well the
model finds negative cases), noting FPR = 1 - specificity, so "the ROC curve visualizes the
relationship between sensitivity and 1-specificity." Just as with precision/recall, the
*threshold* can be adjusted to trade off TPR against FPR (Figure 4.5, p. 165). A "perfect
model" would have a curve going straight up to the point (0,1) - TPR=1, FPR=0. A model that
guesses randomly produces a diagonal line. The area under the ROC curve, **ROC-AUC**
(p. 166), is a single-number summary: 1.0 is a perfect classifier, 0.5 is equivalent to
random guessing, and below 0.5 is worse than random. Computed in scikit-learn with
`sklearn.metrics.roc_auc_score`. The book cautions that "there are situations where the ROC
curve and the associated ROC-AUC metric can be misleading and other evaluation metrics
should be used instead" - for example with a heavily imbalanced dataset where the negative
class dominates, FPR will always stay low even if the model is bad at finding the positive
class.


### 4. What is the precision-recall tradeoff?


> "This phenomenon, that a higher recall generally leads to lower precision and vice
> versa, is called the precision-recall tradeoff."

**How the tradeoff is controlled in practice (p. 161-162)**

Classification models generally work by estimating the probability that a data point
belongs to a class; by default, the model predicts the positive class if that probability
is ≥ 50%. This cutoff is called the **threshold**, and it can be adjusted:

> "If we increase the threshold, only the cases where the model is more certain get
> classified as positive, meaning we get fewer false positives, which in turn leads to
> precision increasing. At the same time, this leads to us missing positive cases, meaning
> there will be more false negatives, which leads to recall decreasing."
>
> "If we decrease the threshold, we allow more data points to be classified as positive,
> which leads to fewer false negatives and thus increases recall. This also leads to
> negative cases incorrectly being classified as positive, meaning we get more false
> positives, which leads to precision decreasing."


### 5. Explain, in broad terms, how each classification model works (having read its scikit-learn documentation)

#### a) Logistic regression (Section 4.3.1, p. 168-173) - [scikit-learn docs](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)

> "Logistic regression is a model that, despite its name 'regression', is used for
> classification problems. It is a binary classifier, and the model performs
> classification by estimating the probability that a data point belongs to a class."

$$p = \sigma(x) = \frac{1}{1+e^{-x}}$$

#### b) Support vector machines (SVM) (Section 4.3.2, p. 176-190) - [scikit-learn docs](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html)

> "For classification problems we instead want to separate data points belonging to
> different classes, and the goal is therefore instead to find as wide a 'road' as
> possible between data points from the two different classes."
 
#### c) Decision trees (Section 4.3.3, p. 190-195) - [scikit-learn docs](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html)

> "Decision trees can, just like SVM, handle both regression and classification problems.
> A decision tree has the same general structure for both regression and classification
> problems, where the tree consists of nodes."

$$G_i = 1-\sum_{k=1}^{K}p_{i,k}^2$$


#### d) Ensemble learning (Section 4.3.4, p. 195-203) - [VotingClassifier docs](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingClassifier.html), [BaggingClassifier docs](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.BaggingClassifier.html)

> "Ensemble Learning [is] a method where several models are combined with the goal of
> creating a model with better predictive and generalization ability than any individual
> model."

#### e) Random forest (Section 4.3.5, p. 203-205) - [scikit-learn docs](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html)

> "Random forest is an ensemble of decision trees usually created through bagging, although
> pasting also occurs."

#### f) Extra trees (Section 4.3.5, p. 205) - [scikit-learn docs](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.ExtraTreesClassifier.html)

> "If, for each split, in addition to a random group of features, we also choose a random
> value for the threshold, we get the model called Extra Tree (Extremely Randomized
> Trees)."


### 6. What does it mean that we can look at feature importance using tree models such as decision trees or random forest?

> "Using decision trees, random forest or extra tree, we can get an estimate of how
> important different features are by using `feature_importances_`. This gives us numbers,
> where a higher number indicates that the variable is more important. Together, the
> numbers also sum to 1."



## Reasoning questions (Resonemangfrågor)

### 7. The precision-recall tradeoff, and the justice system


**What I would answer Stina and Kalle**

Kalle is right to be cautious. As the book explains (Question 4 above, p. 160-161), a
higher precision generally comes at the cost of a lower recall - the **precision-recall
tradeoff**. Concretely, maximizing precision means raising the classification
**threshold** so the model only predicts the positive class when it is very confident. This
reduces false positives (which is exactly what raises precision), but it also means more
actual positive cases get missed - i.e. more false negatives, which lowers recall (Section
4.2.3, p. 161-162: "if we increase the threshold, only the cases where the model is more
certain get classified as positive [...] this leads to us missing positive cases [...]
which leads to recall decreasing"). Taken to the extreme, "the highest possible precision"
could in principle be achieved by a model that predicts the positive class only for the
single observation it is most confident about (or never at all) - which would make
precision look good on paper while recall collapses toward zero, so precision alone is
never the right thing to optimize in isolation.

**When high precision is desirable**

High precision is valuable when a **false positive** is the costly/undesirable error - i.e.
when incorrectly flagging something as positive causes real harm or wasted effort, and
missing some true positives is comparatively more tolerable. The book's own example
(p. 160) is spam filtering: "If our model has a high precision, it means that when an
email is marked as spam, it is often actually spam - but presumably many spam emails will
also end up in our inbox," i.e. we accept a lower recall (some spam gets through) to avoid
the worse outcome of a false positive - a legitimate email being incorrectly sent to the
spam folder, where the user may never see it.

**When high precision can be bad**

High precision (at the cost of recall) is a bad trade when a **false negative** is the
costly/dangerous error - i.e. when *missing* a true positive case is much worse than
incorrectly flagging a negative one. The book's own example (p. 161) is disease screening
for a serious, life-threatening illness: "for serious diseases that can be life-threatening,
it is important that we have a very high recall" - if the model only flags patients it is
extremely confident about (high precision), it will inevitably miss some patients who are
actually sick, and those missed diagnoses could be fatal. In that setting, we deliberately
accept lower precision (more healthy patients flagged for extra tests) in exchange for
higher recall (catching as many actually-sick patients as possible).

**The justice system**

Applying the same precision-recall framework to a criminal court: think of the "positive"
class as *guilty* (i.e. a conviction), so:
- **False positive (FP)** = convicting an innocent person (a wrongful conviction).
- **False negative (FN)** = acquitting a guilty person (a guilty person goes free).

$$\text{Precision} = \frac{TP}{TP+FP} \qquad \text{Recall} = \frac{TP}{TP+FN}$$

A justice system that prioritizes **high precision** for convictions is one where a
person should only be found guilty when the court is very confident (a high "threshold" of
evidence, e.g. "beyond reasonable doubt") - which minimizes the false positives (wrongful
convictions) but, per the precision-recall tradeoff, necessarily also lets more guilty
people go free (more false negatives, i.e. lower recall). This is precisely the ethical
and legal principle usually summarized as "it is better that ten guilty persons escape
than that one innocent suffer" (Blackstone's ratio) - it is a deliberate, explicit choice
to favor precision over recall, because a false positive (imprisoning an innocent person)
is judged to be a far more severe error than a false negative (a guilty person not being
convicted). This mirrors the book's own disease example, but with the asymmetry reversed:
in medicine, the book argues for prioritizing recall (a missed diagnosis can be fatal),
whereas in criminal justice, the near-universal standard is to prioritize precision (a
wrongful conviction is considered a far graver harm than a guilty person going
unpunished). Both are examples of the same underlying principle from the book: there is no
universally "correct" balance between precision and recall - "one needs to analyze the
situation one is making decisions about" (p. 161) to decide which type of error is more
costly, and set the threshold accordingly.


### 8. Explain how Figure 4.8 on page 175 can be interpreted.

**Figure 4.8** (p. 175) shows the decision boundaries of a trained logistic regression
model on the `make_moons` dataset. Its caption states:

> "Decision boundaries for a logistic regression model that classifies observations into
> one of two classes (0/purple or 1/yellow) based on the variables x1 and x2. The
> background color shows the probability (between 0 and 1) that a point belongs to class 1,
> and the color scale is graded on the right."

The interpretation:

- **The dots** are the actual data points from the `make_moons` dataset, colored by their
  true class (dark purple = class 0, yellow = class 1) - the two classes form two
  overlapping crescent/half-moon shapes (introduced on p. 167-168).
- **The background color** at every point in the $(x_1, x_2)$ plane shows the logistic
  regression model's estimated **probability** that a point at that location belongs to
  class 1 (the sigmoid output $p$ from Question 5a), following the color scale on the
  right: dark purple ≈ 0.0 (very likely class 0) through green/teal (uncertain, around
  0.5) to yellow ≈ 1.0 (very likely class 1).
- **The decision boundary** itself is the line where the predicted probability crosses the
  default 50% threshold (Question 4) - i.e. the boundary between the purple-shaded region
  and the yellow-shaded region. Since logistic regression is a **linear model**, this
  boundary is a straight line (visible as the diagonal line where the color bands are
  narrowest/steepest), which is why the text right after the figure states "Logistic
  regression is a linear model, and thus the decision boundaries are also linear, as we
  see in Figure 4.8" (p. 174).

Because the underlying `make_moons` data is **not** linearly separable (the two
crescents interleave), a single straight decision boundary inevitably misclassifies a
substantial number of points "in the middle" where the two classes overlap - which the
book explicitly notes, and which is exactly the motivation for the next figure (4.9),
where polynomial features are added to the logistic regression model to obtain a
**non-linear** decision boundary that fits the crescent shapes far better.


### 9. On page 209 it says "on the training data we use `.fit_transform()`, on the validation data and test data we only use `.transform()`." Explain the logic behind this.

This appears in Section 4.4.1 ("Code Example 1 - Classification of MNIST", p. 209), right
after the code that scales the MNIST pixel data with `StandardScaler`:

```python
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)
X_test = scaler.transform(X_test)
```

**The logic**

`StandardScaler` works by learning two statistics from data - the mean and the standard
deviation of each feature - and then using them to rescale every value (subtract the mean,
divide by the standard deviation). `.fit_transform()` does two things at once: it *fits*
the scaler (i.e. it calculates and stores the mean and standard deviation) **and** it
transforms (rescales) the data it was given. `.transform()` alone only does the second
part - it rescales data using statistics that were *already* fitted earlier; it does not
recalculate them.

The reason the training data gets `.fit_transform()` while the validation and test data
only get `.transform()` is to avoid **data leakage** and to correctly simulate how the
model will be used in the real world. The mean and standard deviation used for scaling must
come **only** from the training data, because:

1. **The validation and test sets are meant to represent unseen, future data.** In a real
   deployment, the model will encounter new observations one at a time (or in small
   batches) - it will never have access to a large batch of "future" data from which to
   compute a fresh mean/standard deviation. If we let the scaler *re-fit* itself on the
   validation or test set, information about those sets (their own mean and spread) would
   leak into the preprocessing step, giving an overly optimistic - and unrealistic -
   picture of how well the model generalizes.
2. **Consistency.** Every observation, whether during training or at prediction time, needs
   to go through *exactly the same transformation* for the model's learned parameters
   (which were fit to the scale of the training data) to remain valid. If the validation or
   test set were scaled using their *own* mean and standard deviation instead of the
   training set's, the same raw pixel value could get mapped to two different scaled
   values depending on which dataset it happened to be in - breaking the correspondence the
   model was trained on.

So the rule is: **fit the scaler once, on the training data only** (`fit_transform`), and
then **reuse that exact same fitted scaler** to transform the validation set, the test set,
and (later) any brand-new data the model is asked to predict on (`transform` only, never
re-fit). This is the same general principle behind why the train/validation/test split
itself exists (Chapters 1-2): the validation and test sets must stay a faithful stand-in
for genuinely new, unseen data, and any step that "peeks" at their statistics before
prediction time - including feature scaling - would undermine that.


In [3]:
### 11. Explain what the code below does, and interpret the result.

from sklearn.metrics import classification_report

y_true = [0, 1, 2, 2, 2]
y_pred = [0, 0, 2, 2, 1]
target_names = ['class 0', 'class 1', 'class 2']
print(classification_report(y_true, y_pred, target_names=target_names))

              precision    recall  f1-score   support

     class 0       0.50      1.00      0.67         1
     class 1       0.00      0.00      0.00         1
     class 2       1.00      0.67      0.80         3

    accuracy                           0.60         5
   macro avg       0.50      0.56      0.49         5
weighted avg       0.70      0.60      0.61         5




This is precisely the worked example the book itself uses to introduce
`classification_report` (Section 4.2.4, p. 163): "the code below demonstrates
`classification_report` from scikit-learn, where precision, recall and f1-score are
printed. The reader is encouraged to read the documentation to understand what everything
means in the printed table."

**What the code does**

`sklearn.metrics.classification_report` takes the true labels (`y_true`) and the model's
predicted labels (`y_pred`) for a **multiclass** classification problem (here, 3 classes:
0, 1 and 2, with `target_names` supplying human-readable labels for each) and computes
**precision**, **recall** and **F1-score** (Question 3c-e) *separately for each class*,
treating that class as the "positive" class and all other classes as "negative" for that
row's calculation - plus a `support` column (how many true instances of that class are in
`y_true`), and finally the overall `accuracy` plus two averaging strategies across classes
(`macro avg` - a plain, unweighted average across classes; `weighted avg` - averaged
weighted by each class's support).

**The result**

```
              precision    recall  f1-score   support

     class 0       0.50      1.00      0.67         1
     class 1       0.00      0.00      0.00         1
     class 2       1.00      0.67      0.80         3

    accuracy                           0.60         5
   macro avg       0.50      0.56      0.49         5
weighted avg       0.70      0.60      0.61         5
```

Interpreting it against the raw data (`y_true = [0, 1, 2, 2, 2]`, `y_pred = [0, 0, 2, 2,
1]`), comparing element by element:

| index | true | predicted | outcome |
|---|---|---|---|
| 0 | 0 | 0 | correct |
| 1 | 1 | 0 | wrong (predicted class 0 instead) |
| 2 | 2 | 2 | correct |
| 3 | 2 | 2 | correct |
| 4 | 2 | 1 | wrong (predicted class 1 instead) |

- **Class 0**: there is 1 true instance (`support=1`) and it *was* predicted correctly, but
  the model also predicted "0" once more for something that was actually class 1 (index
  1) - so out of the 2 times the model predicted class 0, only 1 was correct:
  precision = 1/2 = **0.50**. But since the single true class-0 instance was found,
  recall = 1/1 = **1.00**.
- **Class 1**: there is 1 true instance (`support=1`, at index 4), and the model never
  predicted class 1 for it (it predicted class 2 instead) - so recall = 0/1 = **0.00**.
  The model did predict "1" once (for index 4, incorrectly), and that prediction was wrong,
  so precision = 0/1 = **0.00**. With both precision and recall at 0, F1 is undefined by
  the formula (0/0) and scikit-learn reports it as **0.00** (with a warning in real usage).
- **Class 2**: there are 3 true instances (`support=3`, indices 2, 3, 4), of which 2 were
  correctly predicted as class 2 (indices 2 and 3) and one was missed (index 4, predicted
  as class 1 instead) - so recall = 2/3 ≈ **0.67**. The model predicted "2" exactly twice,
  and both were correct, so precision = 2/2 = **1.00**.
- **accuracy = 0.60**: 3 out of the 5 predictions overall were correct (indices 0, 2, 3),
  which matches Question 3b's formula applied directly to all 5 observations (irrespective
  of class).
- **macro avg**: the plain average across the three classes, treating each class equally
  regardless of how many instances it has, e.g. precision = (0.50+0.00+1.00)/3 ≈ **0.50**.
- **weighted avg**: the same idea, but each class's score is weighted by its `support`
  (so class 2, with 3 of the 5 instances, counts three times as much as classes 0 or 1),
  e.g. precision = (0.50×1 + 0.00×1 + 1.00×3)/5 = **0.70**.

The overall takeaway is that this small example nicely shows why looking only at
`accuracy` (0.60, which sounds "OK") can hide very different behavior *per class* -
class 1 was never correctly identified at all (recall 0.00), which only becomes visible by
inspecting the per-class rows of the classification report.
